source/tutorials/module-system/module-system.md

```diff
@@ -0,0 +1,1027 @@
+```
+
+TODO: Create derivations to get these commands
+
+## A new list option: Declaring generate.requestParams
+
+Let's introduce another option, generate.requestParams. For its type,
+we'll use `listOf <nestedType>`, which is a generic list type where each
+element has to match the given nested type. In our case we want `str` to
+be the nested type, which is a generic string type.
+
+Note that the difference between `str` and `lines` is in their merging
+behavior:
+- For `lines`, multiple definitions get merged by concatenation with
+  newlines
+- For `str`, multiple definitions are not allowed. Which in this case is
+  mostly irrelevant however, since we can't really define a list element
+  multiple times.
+
+```diff
+diff --git a/default.nix b/default.nix
+index 24f9c34..fd5027a 100644
+--- a/default.nix
++++ b/default.nix
+@@ -4,12 +4,21 @@
+     generate.script = lib.mkOption {
+       type = lib.types.lines;
+     };
++
++    generate.requestParams = lib.mkOption {
++      type = lib.types.listOf lib.types.str;
++    };
+   };
+ 
+   config = {
+     generate.script = ''
+       map size=640x640 scale=2 | icat
+     '';
++
++    generate.requestParams = [
++      "size=640x640"
++      "scale=2"
++    ];
+   };
+   
+ }
+```
+
+## Dependencies between options: Using generate.requestParams
+
+A collection of modules generally only has a single option that is meant
+to be evaluated. That's the option that generates the final result we're
+interested in, which in our case is `generate.script`.
+
+In order to build up abstractions on that, we have the ability for
+options to depend on other options. In this case we want the
+`generate.script` option to use the values of `generate.requestParams`.
+We can access the values of options by adding the `config` argument to
+the argument list at the top and using e.g.
+`config.generate.requestParams` to access that options value.
+
+We're then using `lib.concatStringsSep " "` to join each list element
+of the option together into an argument list.
+
+```diff
+diff --git a/default.nix b/default.nix
+index fd5027a..050a98e 100644
+--- a/default.nix
++++ b/default.nix
+@@ -1,4 +1,4 @@
+-{ lib, ... }: {
++{ lib, config, ... }: {
+ 
+   options = {
+     generate.script = lib.mkOption {
+@@ -12,7 +12,9 @@
+ 
+   config = {
+     generate.script = ''
+-      map size=640x640 scale=2 | icat
++      map ${lib.concatStringsSep " "
++            config.generate.requestParams
++           } | icat
+     '';
+ 
+     generate.requestParams = [
+```
+
+## Conditional definitions: Introducing map.zoom option
+
+We now introduce the new option `map.zoom` in order to control the zoom
+level of the map. We'll use a new type for it, `nullOr <type>`, which
+accepts the value `null`, but also values of its argument type. We're
+using `null` here to mean an inferred zoom level.
+
+For this option, we'll set a default value which should be used if it's
+not defined otherwise, which we can do using `mkOption`'s `default`
```

Comment:
This isn't specific to this tutorial, but we need some kind of terminology for the attributes that make up an attribute set passed as an argument. For example, it's incorrect to say that `buildInputs` is an argument to `mkDerivation` since `mkDerivation` is a function that takes a single argument and that argument is an attrset. Maybe something like "input attribute"?

Reply:
I suggest splitting that out into an issue. (GitHub apparently has a UI for that in the `...` menu on the comment.)